File: nodes/core/io/asset_symbol_input_node.py
from typing import Dict, Any
from core.types_registry import get_type
from nodes.base.base_node import BaseNode
from core.types_registry import AssetSymbol, AssetClass, InstrumentType


class AssetSymbolInputNode(BaseNode):
    """Node to create a single AssetSymbol from user parameters."""
    inputs = {}
    outputs = {"symbol": get_type("AssetSymbol")}
    default_params = {
        "ticker": "",
        "asset_class": AssetClass.CRYPTO.name,
        "quote_currency": "USDT",
        "instrument_type": InstrumentType.PERPETUAL.name,
    }
    params_meta = [
        {"name": "ticker", "type": "text", "default": ""},
        {"name": "asset_class", "type": "combo", "default": AssetClass.CRYPTO.name, "options": [e.name for e in AssetClass]},
        {"name": "quote_currency", "type": "text", "default": "USDT"},
        {"name": "instrument_type", "type": "combo", "default": InstrumentType.PERPETUAL.name, "options": [e.name for e in InstrumentType]},
    ]

    async def _execute_impl(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
        # Coerce params to enums and normalized cases
        ticker_value = (self.params.get("ticker") or "").upper()
        asset_class_param = self.params.get("asset_class", AssetClass.CRYPTO.name)
        instrument_type_param = self.params.get("instrument_type", InstrumentType.SPOT.name)
        quote_currency_value = (self.params.get("quote_currency") or None)
        if quote_currency_value:
            quote_currency_value = quote_currency_value.upper()

        # Accept both enum instance and name string for asset_class
        if isinstance(asset_class_param, AssetClass):
            asset_class_value = asset_class_param
        else:
            try:
                asset_class_value = AssetClass[str(asset_class_param).upper()]
            except Exception as e:
                raise ValueError(f"Invalid asset_class: {asset_class_param}") from e

        # Accept both enum instance and name string for instrument_type
        if isinstance(instrument_type_param, InstrumentType):
            instrument_type_value = instrument_type_param
        else:
            try:
                instrument_type_value = InstrumentType[str(instrument_type_param).upper()]
            except Exception as e:
                raise ValueError(f"Invalid instrument_type: {instrument_type_param}") from e

        symbol = AssetSymbol(
            ticker=ticker_value,
            asset_class=asset_class_value,
            quote_currency=quote_currency_value,
            instrument_type=instrument_type_value,
        )
        return {"symbol": symbol}

File: packages/core/src/nodes/core/io/asset-symbol-input-node.ts
// src/nodes/core/io/asset-symbol-input-node.ts
// Translated from: nodes/core/io/asset_symbol_input_node.py

import { Base } from '../../base/base-node';
import {
  AssetClass,
  AssetSymbol,
  InstrumentType,
  getType,
} from '../../../core/types';
import type { NodeInputs, NodeOutputs, ParamMeta, DefaultParams, NodeUIConfig } from '../../../core/types';

/**
 * Node to create a single AssetSymbol from user parameters.
 */
export class AssetSymbolInput extends Base {
  static override inputs: Record<string, unknown> = {};
  static override outputs: Record<string, unknown> = { symbol: getType('AssetSymbol') };

  static uiConfig: NodeUIConfig = {
    size: [240, 140],
    displayResults: false,
    resizable: false,
  };

  static override defaultParams: DefaultParams = {
    ticker: '',
    asset_class: AssetClass.CRYPTO,
    quote_currency: 'USDT',
    instrument_type: InstrumentType.PERPETUAL,
  };

  static override paramsMeta: ParamMeta[] = [
    { name: 'ticker', type: 'text', default: '' },
    {
      name: 'asset_class',
      type: 'combo',
      default: AssetClass.CRYPTO,
      options: Object.values(AssetClass),
    },
    {
      name: 'quote_currency',
      type: 'combo',
      default: 'USD',
      options: ['USD', 'USDC', 'USDT'],
    },
    {
      name: 'instrument_type',
      type: 'combo',
      default: InstrumentType.PERPETUAL,
      options: Object.values(InstrumentType),
    },
  ];

  protected override async executeImpl(_inputs: NodeInputs): Promise<NodeOutputs> {
    // Coerce params to enums and normalized cases
    const tickerValue = String(this.params.ticker ?? '').toUpperCase();
    const assetClassParam = this.params.asset_class ?? AssetClass.CRYPTO;
    const instrumentTypeParam = this.params.instrument_type ?? InstrumentType.SPOT;
    let quoteCurrencyValue = this.params.quote_currency ?? null;

    if (quoteCurrencyValue) {
      quoteCurrencyValue = String(quoteCurrencyValue).toUpperCase();
    }

    // Parse asset_class - accept both enum instance and name string
    let assetClassValue: AssetClass;
    if (Object.values(AssetClass).includes(assetClassParam as AssetClass)) {
      assetClassValue = assetClassParam as AssetClass;
    } else {
      const upperParam = String(assetClassParam).toUpperCase();
      if (Object.values(AssetClass).includes(upperParam as AssetClass)) {
        assetClassValue = upperParam as AssetClass;
      } else {
        throw new Error(`Invalid asset_class: ${assetClassParam}`);
      }
    }

    // Parse instrument_type - accept both enum instance and name string
    let instrumentTypeValue: InstrumentType;
    if (Object.values(InstrumentType).includes(instrumentTypeParam as InstrumentType)) {
      instrumentTypeValue = instrumentTypeParam as InstrumentType;
    } else {
      const upperParam = String(instrumentTypeParam).toUpperCase();
      if (Object.values(InstrumentType).includes(upperParam as InstrumentType)) {
        instrumentTypeValue = upperParam as InstrumentType;
      } else {
        throw new Error(`Invalid instrument_type: ${instrumentTypeParam}`);
      }
    }

    const symbol = new AssetSymbol(
      tickerValue,
      assetClassValue,
      quoteCurrencyValue ? String(quoteCurrencyValue) : undefined,
      instrumentTypeValue
    );

    return { symbol };
  }
}
